File: app/services/weather_service.py
"""
Mock weather service (dependency-free).

Mirrors src/services/weatherService.js.
"""

from __future__ import annotations

from dataclasses import asdict, dataclass
from datetime import date, datetime, timedelta
from typing import Literal


Units = Literal["metric", "imperial"]


def _compute_mock_temperature(city: str, units: Units) -> int | float:
    """Deterministic-ish mock temperature based on city name."""
    base_c = 12
    variance = sum(ord(ch) for ch in city.lower())
    temp_c = base_c + (variance % 15)  # 12..26

    if units == "imperial":
        return round((temp_c * 9) / 5 + 32)
    return temp_c


def _compute_mock_description(seed: str) -> str:
    """Return a simple mock weather description."""
    options = ["Clear sky", "Partly cloudy", "Overcast", "Light rain", "Breezy"]
    idx = sum(ord(ch) for ch in seed.lower()) % len(options)
    return options[idx]


@dataclass(frozen=True)
class MockWeather:
    city: str
    units: Units
    temperature: int | float
    description: str
    humidity: int
    windSpeed: int | float
    timestamp: str


@dataclass(frozen=True)
class MockForecastDay:
    date: str
    temperature: int | float
    description: str
    humidity: int
    windSpeed: int | float


# PUBLIC_INTERFACE
def get_mock_weather(*, city: str, units: Units) -> dict:
    """Get mock current weather payload for a city."""
    temperature = _compute_mock_temperature(city, units)

    humidity = 40 + (len(city) * 7) % 55  # 40..94
    wind_speed_metric = 2 + (len(city) % 6)  # 2..7 m/s
    wind_speed = round(wind_speed_metric * 2.237) if units == "imperial" else wind_speed_metric

    payload = MockWeather(
        city=city,
        units=units,
        temperature=temperature,
        description=_compute_mock_description(city),
        humidity=humidity,
        windSpeed=wind_speed,
        timestamp=datetime.utcnow().isoformat(timespec="seconds") + "Z",
    )
    return asdict(payload)


# PUBLIC_INTERFACE
def get_mock_forecast(*, city: str, days: int) -> dict:
    """Get mock multi-day forecast payload for a city."""
    base = get_mock_weather(city=city, units="metric")

    start = date.today()
    forecast: list[dict] = []

    for i in range(days):
        d = start + timedelta(days=i + 1)

        temp_c = base["temperature"] + ((1 if i % 2 == 0 else -1) * (i + 1))
        humidity = max(25, min(95, base["humidity"] + (5 if i % 3 == 0 else -3)))
        wind_speed = max(1, base["windSpeed"] + (1 if i % 2 == 0 else -1))

        day_payload = MockForecastDay(
            date=d.isoformat(),
            temperature=temp_c,
            description=_compute_mock_description(f"{city}-{i}"),
            humidity=humidity,
            windSpeed=wind_speed,
        )
        forecast.append(asdict(day_payload))

    return {"city": city, "days": days, "forecast": forecast}

File: src/services/weatherService.js
/**
 * Mock weather service.
 *
 * This layer is intentionally dependency-free and does not call external APIs yet.
 * Later, it can be replaced by a real provider integration (e.g., OpenWeather).
 */

/**
 * Deterministic-ish mock temperature based on city name so different cities look different.
 * @param {string} city
 * @param {"metric"|"imperial"} units
 * @returns {number}
 */
function computeMockTemperature(city, units) {
  const baseC = 12;
  const variance = city
    .toLowerCase()
    .split("")
    .reduce((acc, ch) => acc + ch.charCodeAt(0), 0);

  const tempC = baseC + (variance % 15); // 12..26
  if (units === "imperial") {
    return Math.round((tempC * 9) / 5 + 32);
  }
  return tempC;
}

/**
 * Return a simple mock weather description.
 * @param {string} city
 * @returns {string}
 */
function computeMockDescription(city) {
  const options = ["Clear sky", "Partly cloudy", "Overcast", "Light rain", "Breezy"];
  const idx =
    city
      .toLowerCase()
      .split("")
      .reduce((acc, ch) => acc + ch.charCodeAt(0), 0) % options.length;
  return options[idx];
}

/**
 * @param {object} input
 * @param {string} input.city
 * @param {"metric"|"imperial"} input.units
 * @returns {{city:string, units:"metric"|"imperial", temperature:number, description:string, humidity:number, windSpeed:number, timestamp:string}}
 */
// PUBLIC_INTERFACE
function getMockWeather({ city, units }) {
  /** Get mock current weather payload for a city. */
  const temperature = computeMockTemperature(city, units);

  const humidity = 40 + (city.length * 7) % 55; // 40..94
  const windSpeedMetric = 2 + (city.length % 6); // 2..7 m/s
  const windSpeed = units === "imperial" ? Math.round(windSpeedMetric * 2.237) : windSpeedMetric;

  return {
    city,
    units,
    temperature,
    description: computeMockDescription(city),
    humidity,
    windSpeed,
    timestamp: new Date().toISOString()
  };
}

/**
 * @param {object} input
 * @param {string} input.city
 * @param {number} input.days
 * @returns {{city:string, days:number, forecast:Array<{date:string, temperature:number, description:string, humidity:number, windSpeed:number}>}}
 */
// PUBLIC_INTERFACE
function getMockForecast({ city, days }) {
  /** Get mock multi-day forecast payload for a city. */
  const today = new Date();
  const base = getMockWeather({ city, units: "metric" });

  const forecast = Array.from({ length: days }, (_, i) => {
    const date = new Date(today);
    date.setDate(today.getDate() + i + 1);

    // Simple variation day-by-day
    const tempC = base.temperature + ((i % 2 === 0 ? 1 : -1) * (i + 1));
    const humidity = Math.max(25, Math.min(95, base.humidity + (i % 3 === 0 ? 5 : -3)));
    const windSpeed = Math.max(1, base.windSpeed + (i % 2 === 0 ? 1 : -1));

    return {
      date: date.toISOString().slice(0, 10),
      temperature: tempC,
      description: computeMockDescription(`${city}-${i}`),
      humidity,
      windSpeed
    };
  });

  return {
    city,
    days,
    forecast
  };
}

module.exports = {
  getMockWeather,
  getMockForecast
};
